src/snowflake/cli/api/metrics.py

```diff
@@ -86,7 +207,109 @@ def increment_counter(self, name: str, value: int = 1) -> None:
+        E.g. through different scopes/files for when a context manager is not enough
+        Requires a corresponding call to end_step with the (optional) exception caught
+
+        :returns: id of the started step which can be passed to end_step
+
+        :raises CliMetricsInvalidUsageError: if the step name is empty
+        """
+        parent_step = self._current_step
+        new_step = _CLIMetricsStep(name, parent_step)
+        self._executing_steps.append(new_step)
+        new_step.start()
+        return new_step.step_id
+
+    def end_step(
+        self,
+        step_id: Optional[int] = None,
+        step_name: Optional[str] = None,
+        error: Optional[BaseException] = None,
+    ) -> None:
+        """
+        Manually ends either the most recently started step or the step with the provided id/name
+
+        Context managers can automatically check if there were errors, but if manually
+        starting/ending steps, then the caller will need to catch and pass the error themselves
+
+        :raises CliMetricsInvalidUsageError: if step could not be found or there are no executing steps
+        """
+        if step_id is not None:
+            found_step = next(
+                (step for step in self._executing_steps if step.step_id == step_id),
+                None,
+            )
+        elif step_name is not None:
+            found_step = next(
+                (
+                    step
+                    for step in reversed(self._executing_steps)
+                    if step.name == step_name
+                ),
+                None,
+            )
+        else:
+            found_step = self._current_step
+
+        if not found_step:
+            if step_id or step_name:
+                raise CLIMetricsInvalidUsageError(
+                    f"step with {'id' if step_id else 'name'} '{step_id or step_name}' could not be ended because it could not be found"
+                )
+            else:
+                raise CLIMetricsInvalidUsageError(
+                    "current step could not be ended because no steps are executing"
+                )
+
+        found_step.end(error)
+        self._executing_steps.remove(found_step)
+        self._finished_steps.append(found_step)
+
+    def flush_steps(self, error: Optional[BaseException] = None) -> None:
+        """
+        Useful if you are using the manual start/end steps and an error
+        propagated up, requiring you to clear out all the executing steps
+        """
+        while self._current_step:
+            self.end_step(error=error)
+
     @property
     def counters(self) -> Dict[str, int]:
         # return a copy of the original dict to avoid mutating the original
         return self._counters.copy()
+
+    @property
+    def steps(self) -> List[Dict[str, Union[str, int, float, None]]]:
+        """
+        returns the finished steps tracked throughout a command, sorted by start time
+        """
```

Comment:
I have no idea what the return type represents here, other than it's a list of dictionaries. Would `List[dict]` suffice? Basically you're returning a list of JSON objects.